ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java

```diff
@@ -301,4 +321,70 @@ private static void registerPassivationStoreTransformers_1_2_1_and_1_3_0(Resourc
                 .setValueConverter(AttributeConverter.Factory.createHardCoded(new ModelNode().set(TimeUnit.SECONDS.name()), true), EJB3SubsystemModel.IDLE_TIMEOUT_UNIT)
         ;
     }
+
+    /**
+     * This Combined Transformer manages this transformation from EAP7 to previous versions (attribute definition/xml markup):
+     *
+     * DEFAULT_SFSB_CACHE / cache-ref --> DEFAULT_CLUSTERED_SFSB_CACHE / clustered-cache-ref
+     * DEFAULT_SFSB_PASSIVATION_DISABLED_CACHE / passivation-disabled-cache-ref --> DEFAULT_SFSB_CACHE / cache-ref
+     *
+     */
+    private static class SfsbChacheRefTransformer implements CombinedTransformer {
+        private final Map<String, String> renames;
+
+        public SfsbChacheRefTransformer(){
+            renames = new HashMap<>();
+            renames.put(DEFAULT_SFSB_CACHE, EJB3SubsystemModel.DEFAULT_CLUSTERED_SFSB_CACHE);
+            renames.put(DEFAULT_SFSB_PASSIVATION_DISABLED_CACHE, EJB3SubsystemModel.DEFAULT_SFSB_CACHE);
+        }
+
+        @Override
+        public TransformedOperation transformOperation(TransformationContext context, PathAddress address, ModelNode operation) throws OperationFailedException {
+            if (operation != null && !(operation.hasDefined(OPERATION_HEADERS) && operation.get(OPERATION_HEADERS, "push-to-servers").asBoolean(false)) ){
```

Comment:
I cannot evaluate if the use of this header is the correct solution. There is a corner case with the transformation required for this issue. What we have to do is implement the following transformation:
cache-ref --> clustered-cache-ref
passivation-disabled-cache-ref --> cache-ref

Here, the custom operation transformer is invoked more than once for different addresses to transform the same operation. There are two code paths which trigger this request of transformation:

1. Operations pushed into the HC
    https://github.com/wildfly/wildfly-core/blob/46ad87495026705fd559698d32069978b38f7071/host-controller/src/main/java/org/jboss/as/domain/controller/operations/coordination/HostControllerUpdateTask.java#L97

2. Operations pushed into the Server groups pushed by the HC
    https://github.com/wildfly/wildfly-core/blob/46ad87495026705fd559698d32069978b38f7071/host-controller/src/main/java/org/jboss/as/domain/controller/operations/coordination/DomainRolloutStepHandler.java#L355

The operation transformation triggered by 2) is invoked for an operation already transformed by the invocation 1). That situation is the corner case for this issue because if we are transforming `passivation-disabled-cache-ref` into `cache-ref`, we will end up transforming again from `cache-ref` to `clustered-cache-ref` due to 2).

The presence of the header "push-to-servers" in the invocation 2) seems to solve this dilemma, but I'm not sure if this is the correct solution.

Reply:
I think this is an ok workaround. I filed https://issues.jboss.org/browse/WFCORE-3439 to look into why this double transformation happens.